nodejs/awsx/utils.ts

```diff
@@ -103,3 +105,33 @@ export function hasTrueBooleanMember(obj: any, memberName: string | number | sym
 
     return val === true;
 }
+
+/** @internal */
+export function promiseResult<T>(promise: Promise<T>): T {
+    enum State {
+        running,
+        finishedSuccessfully,
+        finishedWithError,
+    }
+
+    let result: T;
+    let error = undefined;
+    let state = <State>State.running;
+
+    promise.then(
+        val => {
+            result = val;
+            state = State.finishedSuccessfully;
+        },
+        err => {
+            error = err;
+            state = State.finishedWithError;
+        });
+
+    deasync.loopWhile(() => state === State.running);
+    if (state === State.finishedWithError) {
+        throw error;
+    }
```

Comment:
extremely simple async-blocking.  The promise above runs, and when complete sets result/error/state.  `loopWhile` then just pumps the node main loop as long as the condition inside keeps returning true.  Code for deasync is extremely trivial:

`loopWhile` is:

https://github.com/abbr/deasync/blob/90e24d8206b6d8d8ada9179ff00be5acf245c68a/index.js#L68-L73

And the actual `Run` function it calls into is:

https://github.com/abbr/deasync/blob/90e24d8206b6d8d8ada9179ff00be5acf245c68a/src/deasync.cc#L8-L13

Reply:
While this is risky to take this dependency on this native code module, it feels like this is so minimal and so targeted, that it should be an acceptable risk given the benefit we get.

Ideally, nodejs grpc could just expose sync calls.  But, in the absence of that, this gets us to that state at a safe enough level IMO.

Reply:
Note: the intent would be to move this down to the provider level.  So `pulumi/aws` would expose two methods `getAvailabilityZones(): Promise<GetAvailabilityZonesResult>` and `getAvailabilityZonesSync(): GetAvailabilityZonesResult`.  We'd keep the async versions around for compat (and because you should prefer them in cases where you can avoid the sync).  

I also thought about just making the existing calls return the non-promise result.  THat would be non-breaking for people who are calling `await` today.  But it would break people who explicit do things like `.then(...)`.